plugins/renepay/flow.c

```diff
@@ -0,0 +1,849 @@
+#include "config.h"
+#include <assert.h>
+#include <ccan/asort/asort.h>
+#include <ccan/tal/str/str.h>
+#include <ccan/tal/tal.h>
+#include <common/type_to_string.h>
+#include <math.h>
+#include <plugins/renepay/debug.h>
+#include <plugins/renepay/flow.h>
+#include <stdio.h>
+
+#ifndef SUPERVERBOSE
+#define SUPERVERBOSE(...)
+#else
+#define SUPERVERBOSE_ENABLED 1
+#endif
+
+#define MAX(x, y) (((x) > (y)) ? (x) : (y))
+#define MIN(x, y) (((x) < (y)) ? (x) : (y))
+
+bool chan_extra_is_busy(struct chan_extra const * const ce)
+{
+	if(ce==NULL)return false;
+	return ce->half[0].num_htlcs || ce->half[1].num_htlcs;
+}
+
+const char *fmt_chan_extra_map(
+		const tal_t *ctx,
+		struct chan_extra_map* chan_extra_map)
+{
+	tal_t *this_ctx = tal(ctx,tal_t);
+	char *buff = tal_fmt(ctx,"Uncertainty network:\n");
+	struct chan_extra_map_iter it;
+	for(struct chan_extra *ch = chan_extra_map_first(chan_extra_map,&it);
+	    ch;
+	    ch=chan_extra_map_next(chan_extra_map,&it))
+	{
+		const char *scid_str =
+			type_to_string(this_ctx,struct short_channel_id,&ch->scid);
+		for(int dir=0;dir<2;++dir)
+		{
+			tal_append_fmt(&buff,"%s[%d]:(%s,%s)\n",scid_str,dir,
+				type_to_string(this_ctx,struct amount_msat,&ch->half[dir].known_min),
+				type_to_string(this_ctx,struct amount_msat,&ch->half[dir].known_max));
+		}
+	}
+	tal_free(this_ctx);
+	return buff;
+}
+
+struct chan_extra *new_chan_extra(
+		struct chan_extra_map *chan_extra_map,
+		const struct short_channel_id scid,
+		struct amount_msat capacity)
+{
+	struct chan_extra *ce = tal(chan_extra_map, struct chan_extra);
+
+	ce->scid = scid;
+	ce->capacity=capacity;
+	for (size_t i = 0; i <= 1; i++) {
+		ce->half[i].num_htlcs = 0;
+		ce->half[i].htlc_total = AMOUNT_MSAT(0);
+		ce->half[i].known_min = AMOUNT_MSAT(0);
+		ce->half[i].known_max = capacity;
+	}
+	chan_extra_map_add(chan_extra_map, ce);
+
+	/* Remove self from map when done */
+	// TODO(eduardo):
+	// Is this desctructor really necessary? the chan_extra will deallocated
+	// when the chan_extra_map is freed. Anyways valgrind complains that the
+	// hash table is removing the element with a freed pointer.
+	// tal_add_destructor2(ce, destroy_chan_extra, chan_extra_map);
```

Comment:
If you free the "ce" though, the pointer stays in the hash table!  valgrind is right, this can be fatal.

You have two choices:
1.  Remove self from table on free automatically, using destructor.  This is simplest!
2. Remove from table manually.  This optimization might happen after benchmarking, only!